add plugin client factory

fbourigault opened the pull request:
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | partially https://github.com/php-http/HttplugBundle/issues/109
| Documentation   | TODO
| License         | MIT

This is required to display plugins created by 3rd party libraries in the HttplugBundle profiler. See https://github.com/php-http/HttplugBundle/issues/109#issuecomment-300461535 for all puzzle pieces.

#### To Do

- [x] Update the changelog.
- [ ] Update the documentation.

Nyholm commented on src/PluginClientFactory.php:
```diff
@@ -0,0 +1,19 @@
+namespace Http\Client\Common;
+
+/**
+ * Create a PluginClient.
```
We could have a more expressive comment.

Nyholm replied:
Or maybe remove this class comment and only use the one on the interface.

Nyholm commented on src/PluginClientFactoryInterface.php:
```diff
@@ -0,0 +1,25 @@
+use Http\Client\HttpClient;
+
+/**
+ * Create a PluginClient.
```
Same here. :)

fbourigault replied:
I'm open to any suggestion :)

Nyholm replied:
Maybe: 

> Create a PluginClient object loaded with an array of plugins. A factory may add or modify the plugins or change the HttpClient.

Nyholm commented on src/PluginClientFactoryInterface.php:
```diff
@@ -0,0 +1,25 @@
+     * @param Plugin[]                   $plugins
+     * @param array                      $options
+     *
+     * @see PluginClient constructor for $options details.
```
No, we cannot "see the PluginClient". We are not aware of any client implementation. We should write something like "array of options that are passed to the client."

fbourigault replied:
I don't think so. There is only one PluginClient without a dedicated interface and this implementation is final. There is no way to write an other PluginClient!

Nyholm replied:
You are correct. I had in mind that one could create ANY `PluginClient` with this. But the `PluginClientFactoryInterface` creates an instance of the concrete class `PluginClient`. That is a huge different from our other `*FactoryInterface`s. They create an instance of an interface.

Nyholm commented:
Note: 
The `PluginClientFactoryInterface` creates an instance of the concrete class `PluginClient`. That is a huge different from our other `*FactoryInterface`s. They create an instance of an interface.

sagikazarmark commented:
TBH, I don't really see the point of this interface.

fbourigault commented:
The idea behind is to use discovery to find a concrete PluginClientFactory. When the bundle profiling is enabled, a profile aware PluginClientFactory is decorates each plugin with a ProfilePlugin and add the StackPlugin at the begining. See https://github.com/php-http/HttplugBundle/issues/109#issuecomment-300461535 for a whole plan details.

sagikazarmark commented:
Why do we need a factory to decorate plugins? And why do we have to discover it? I still think it's configuration: when profiling is enabled, hook into DI and wrap plugins, otherwise don't.

fbourigault commented:
It's all about displaying plugins from 3rd party like php-github-api in the profiler. With such implementation, you can integrate any 3rd party library (which use this interface and the discovery) in the profiler without having any specific code in the 3rd party library, without having a glue bundle and without having to write a line of configuration.

fbourigault commented:
I updated the PR to add a note in the changelog. I also adressed @Nyholm comments about docblocks. I added in my to do list the puli stuff. Can someone help me with it?

fbourigault commented:
How can I test puli resources declaration?

fbourigault commented:
This is now updated to provide a solution which does not require `php-http/discovery` to profile 3rd party library clients.

Nyholm commented on src/PluginClientFactory.php:
```diff
@@ -0,0 +1,48 @@
+     * Set the factory to use. The callable to provide must have the same arguments and return type as
+     * PluginClientFactory::createClient.
+     *
+     * @internal
```
Maybe state why?

> This is used by the HTTPlugBundle to provide a better Symfony integration

Nyholm commented on src/PluginClientFactory.php:
```diff
@@ -0,0 +1,48 @@
+     */
+    public function createClient($client, array $plugins = [], array $options = [])
+    {
+        $factory = static::$factory ?: function ($client, array $plugins, array $options) {
```
Why not simply do: 

```php
if (static::$factory) {
  return $factory($client, $plugins, $options);
}

return new PluginClient($client, $plugins, $options);
```

It would be easier to read.

fbourigault commented:
Thank you @Nyholm. I addressed all your remarks.

Nyholm commented on src/PluginClientFactory.php:
```diff
@@ -0,0 +1,49 @@
+    public function createClient($client, array $plugins = [], array $options = [])
+    {
+        if (static::$factory) {
+            return (static::$factory)($client, $plugins, $options);
```
This is not allowed in PHP 5.x.  ( I just found out)

```php
$func = static::$factory;
return $func($client, $plugins, $options);

// Or: (This might work)
return static::$factory($client, $plugins, $options); 
```

fbourigault commented:
Should be fixed now. Please, don't merge until it's ok on other packages.

Nyholm commented:
@fbourigault Is this ready to merge? 

Do @dbu or @sagikazarmark want to have a final review?

sagikazarmark commented on src/PluginClientFactory.php:
```diff
@@ -0,0 +1,51 @@
+    /**
+     * @var callable
+     */
+    private static $factory;
```
Is this the workaround for not having an interface? If so, having an interface for this is still less evil. This solution provides zero type safety (even if the support PHP versions do not support return types).

fbourigault replied:
As the stuff is marked internal, we choosed to avoid over engineering and to not use an interface.

fbourigault commented:
I added the `client_name` option.

Nyholm commented:
I will merge this, but we should review it a final time before releasing. 

Thank you for the comments and thank you for the work on this PR

dbu commented on src/PluginClientFactory.php:
```diff
@@ -0,0 +1,55 @@
+        if (static::$factory) {
+            $factory = static::$factory;
+
+            return $factory($client, $plugins, $options);
```
deos simply `return static::$factory(...)` not work?

fbourigault replied:
Only in PHP 7+

dbu replied:
weird. can you add a comment that this is to work with php 5 and can be removed in 7?

fbourigault replied:
I was wrong. It doesn't work in any PHP version: "Function name must be a string".
See https://3v4l.org/WA5ek.

dbu replied:
alright. then maybe a comment that using it directly does not work at least up to php 7.1?

fbourigault replied:
Doesn't work too in 7.2.0alpha2. IMHO it doesn't worth to write a comment about something not supported by the language itself.

dbu commented on src/PluginClientFactory.php:
```diff
@@ -0,0 +1,55 @@
+     *
+     * @return PluginClient
+     */
+    public function createClient($client, array $plugins = [], array $options = [])
```
why is this method dynamic but the factory callable static? this means i must instantiate the factory, but i can only set one single callable. why not make the callable an optional constructor argument?

fbourigault replied:
The setFactory is static as we need to change the implementation at runtime when bundle profiling is enabled and, as we aim to do 0config integration, we can't have something non-static here.

dbu replied:
ah, i see. maybe a bit of comment in the code would help so we remember in the future ;-)

dbu commented:
could we use the class docblock to explain *why* this factory should be used?
